### PR TITLE
feat : Impelment the logic of ExerciseScreen(Easy & hard) and write the relative tests

### PR DESCRIPTION
**Pull request by SaFouNaldo**
## Related Issue
This PR addresses issue #143 .

## Summary of Changes
This PR focuses on enhancing ExerciseScreenEasy and ExerciseScreenHard to automatically progress to the next letter when a correct sign is made. Key changes include:

### 1 - MainActivity.kt:

- Added native permission checks for camera use since both screens use the camera preview.
- Removed button checks as they are no longer needed, improving test coverage slightly.

### 2 -  Utils.kt:

- Added CameraPlaceholder for better readability and replaced deprecated code.
UtilsTest.kt:

- Added a test to verify the camera preview, increasing test coverage to 85%.
### 3 - ExerciseScreenEasy.kt:

- Added gesture handling in a new function named handleGestureMatching.
- Removed the previously hardcoded SuccessButton.
### 4 - ExerciseScreenHard.kt:

- Similar changes as ExerciseScreenEasy, but implemented in a more concise way.
### 5 - ExerciseScreenEasyTest.kt:

- Added tests for the new implementation, resulting in 80% test coverage.
### 6 - ExerciseScreenHardTest.kt:

- Added tests for ExerciseScreenHard, reaching 68% test coverage, which can be significantly improved.
### 7 - strings.xml:

- Added two string arrays for words: one for 3-letter words and another for longer words, improving readability and removing hardcoded values.

## Future Work
- Rename ExerciseScreenHard to ExerciseScreenMedium.
- Create a new screen named ExerciseScreenHard that focuses on forming sentences.
- Improve test coverage for ExerciseScreenEasyTest and ExerciseScreenHardTest.

## Note on PR Size
- I acknowledge that this is a medium-sized PR, containing multiple updates across different parts of the project. These changes were grouped together to maintain consistency and ensure all related functionality is fully integrated and tested. Breaking them into smaller PRs could have led to incomplete features and inconsistent states, so I opted to submit them as a complete set.

- These changes enhance readability, maintainability, test coverage, and overall user experience by enabling seamless automatic letter progression in the exercise screens.

### CODE REVIEW

**Comment by SaFouNaldo** [comment]
As SonarCloud states that there is a duplication code I'll update asap

File: app/src/main/java/com/github/se/signify/ui/Utils.kt [hunk 1]
```diff
@@ -206,3 +229,102 @@ fun getLetterIconResId(letter: Char): Int {
...
   }
 }
+/**
+ * Composable function that serves as a placeholder for the camera preview. This function requests
+ * camera permission and displays the camera preview if permission is granted.
+ *
+ * @param handLandMarkViewModel The ViewModel responsible for managing hand landmark detection.
+ * @param modifier Modifier to be applied to the camera placeholder.
+ */
+@Composable
```
Comment (NBCCF025) [inline app/src/main/java/com/github/se/signify/ui/Utils.kt:239]: This function is a bit long and could use helper functions. For example the box could have been created in other helper functions that you call in your if/else.
It is a best practice to add documentation for general function, espacialy for the one in Utils.kt so that's great.

Overall good job as dealing with the camera is not easy.

File: app/src/main/java/com/github/se/signify/ui/screens/home/ExerciseScreenEasy.kt [hunk 2]
```diff
@@ -200,42 +160,131 @@ fun WordLayer(words: List<String>, currentWordIndex: Int, currentLetterIndex: In
...
-  Box(
-      modifier =
-          modifier
-              .fillMaxWidth()
-              .padding(16.dp)
-              .height(350.dp)
-              .background(colorResource(R.color.black), shape = RoundedCornerShape(16.dp))
-              .border(2.dp, colorResource(R.color.white), shape = RoundedCornerShape(16.dp)),
-      contentAlignment = Alignment.Center) {}
+fun ExerciseScreenCommon(
```
Comment (NBCCF025) [inline app/src/main/java/com/github/se/signify/ui/screens/home/ExerciseScreenEasy.kt:201]: The function ExerciseScreenCommon() could have been created in Utils.kt as it is a common element used in different screen like ExerciseScreenEasy and ExerciseScreenHard.

**Review by NBCCF025 — APPROVED** [review]
This PR is well-organized and addresses important feature enhancements and improvements across multiple files. Here are my observations:

1. **MainActivity.kt**: Introducing native permission checks for the camera is a solid improvement, removing outdated button checks and ensuring permissions are appropriately handled. This should help reduce redundancy in test cases and make them more reliable.

2. **Utils.kt and UtilsTest.kt**: The addition of `CameraPlaceholder` and removal of deprecated code are good steps towards improving readability and future-proofing the code. The enhanced test coverage is also beneficial, reaching 85%, which adds confidence in the stability of the camera preview functionality.

3. **Exercise Screens (Easy and Hard)**: The approach of handling gesture detection in `handleGestureMatching` is an excellent modularization choice, improving readability and maintainability. Removing the hardcoded `SuccessButton` adds flexibility. The specific improvements made in `ExerciseScreenHard` to ensure concise functionality is appreciated. This modular structure should simplify future changes, especially with the planned renaming and screen additions.

4. **Tests**: The new test cases for the updated implementations in `ExerciseScreenEasyTest.kt` and `ExerciseScreenHardTest.kt` are valuable additions, though the coverage in `ExerciseScreenHardTest.kt` (68%) indicates room for further improvement.

5. **strings.xml**: Replacing hardcoded values with arrays for better readability is a positive change, making future modifications simpler and enhancing consistency across the application.

6. **PR Size Justification**: The rationale for grouping these changes into a single PR makes sense given the interrelated nature of the updates, and the comprehensive nature of the PR is reasonable for consistency.

7. **Tests With Emulator**: I tested the camera through the exercise screens and it work pretty well. Except when I have difficulty to do a letter (like T) the app crash after some time. I don't know if the problem come from the emulator or the hand detection algorithm.

Overall, this PR offers a well-rounded update that improves code readability, maintainability, and user experience. Future work, as outlined, will further enhance the quality of these features. Good job for the hard work put in this PR.

**Review by nada-ezziti — APPROVED** [review]
This PR introduces several significant improvements to the ExerciseScreenEasy and ExerciseScreenHard, enhancing the user experience by enabling automatic progression to the next letter when a correct sign is made. It also includes updates to permissions, utilities, tests, and resources to ensure a cohesive implementation.
_______________
Here’s my feedback:

**Automatic Letter Progression**
- The implementation of handleGestureMatching centralizes gesture matching logic, streamlining functionality and improving maintainability.
- The onSuccess function efficiently handles transitions between letters, words, and exercises, significantly enhancing user flow.

**Permissions and Deprecated Code**
- Added native camera permission checks in MainActivity.kt, ensuring a seamless experience across devices.
- Replaced deprecated code in Utils.kt with a new CameraPlaceholder, improving readability and code longevity.

**Testing Enhancements**
Achieved a notable increase in test coverage:
- ExerciseScreenEasyTest.kt now at 80% coverage.
- ExerciseScreenHardTest.kt coverage at 68%, with clear notes for future improvement.
- Introduced a test for camera preview in UtilsTest.kt, raising overall test coverage to 85%.

**Resource Improvements**
- Refactored hardcoded words into string arrays (strings.xml) for better readability and localization support.

**Future-Proofing and Roadmap**
- Provided a clear roadmap for future changes, including renaming ExerciseScreenHard to ExerciseScreenMedium and focusing on forming sentences in a new ExerciseScreenHard.
- Highlighted plans to improve test coverage, which demonstrates a proactive approach to technical debt.
__________
**Design and Implementation Notes**

**Consistency**
- Grouping related changes into a single PR helps maintain consistency across screens, utilities, and tests.
- ExerciseScreenCommon is well-designed for reusability, ensuring maintainability as new screens are added.

**User Experience**
- The seamless transition between letters and words enhances interactivity and reduces user effort.
- Visual improvements, such as the WordLayer composable with emphasized current letters, contribute to a polished UI.

**Testing Coverage**
- While the increase in test coverage is commendable, focusing on the ExerciseScreenHardTest to bring it closer to the standard set by ExerciseScreenEasyTest would be a welcome next step.

**Suggestions for Future Improvements**

- Enhance documentation for key functions such as handleGestureMatching and onSuccess to ensure clarity for all contributors.
- Explore additional edge-case tests, particularly for gesture recognition, to further solidify the robustness of the screens.
- I have the same problem as Nassim for the emulator testing. 
_____
**Overall Rating:  Excellent Work**

This PR is a well-rounded contribution that significantly improves the exercise screens, user experience, and test coverage while laying the groundwork for future enhancements. Thank you for the meticulous effort and thoughtful execution!

**Comment by NBCCF025** [comment]
> This PR is well-organized and addresses important feature enhancements and improvements across multiple files. Here are my observations:
> 
>     1. **MainActivity.kt**: Introducing native permission checks for the camera is a solid improvement, removing outdated button checks and ensuring permissions are appropriately handled. This should help reduce redundancy in test cases and make them more reliable.
> 
>     2. **Utils.kt and UtilsTest.kt**: The addition of `CameraPlaceholder` and removal of deprecated code are good steps towards improving readability and future-proofing the code. The enhanced test coverage is also beneficial, reaching 85%, which adds confidence in the stability of the camera preview functionality.
> 
>     3. **Exercise Screens (Easy and Hard)**: The approach of handling gesture detection in `handleGestureMatching` is an excellent modularization choice, improving readability and maintainability. Removing the hardcoded `SuccessButton` adds flexibility. The specific improvements made in `ExerciseScreenHard` to ensure concise functionality is appreciated. This modular structure should simplify future changes, especially with the planned renaming and screen additions.
> 
>     4. **Tests**: The new test cases for the updated implementations in `ExerciseScreenEasyTest.kt` and `ExerciseScreenHardTest.kt` are valuable additions, though the coverage in `ExerciseScreenHardTest.kt` (68%) indicates room for further improvement.
> 
>     5. **strings.xml**: Replacing hardcoded values with arrays for better readability is a positive change, making future modifications simpler and enhancing consistency across the application.
> 
>     6. **PR Size Justification**: The rationale for grouping these changes into a single PR makes sense given the interrelated nature of the updates, and the comprehensive nature of the PR is reasonable for consistency.
> 
>     7. **Tests With Emulator**: I tested the camera through the exercise screens and it work pretty well. Except when I have difficulty to do a letter (like T) the app crash after some time. I don't know if the problem come from the emulator or the hand detection algorithm.
> 
> 
> Overall, this PR offers a well-rounded update that improves code readability, maintainability, and user experience. Future work, as outlined, will further enhance the quality of these features. Good job for the hard work put in this PR.

The point 7 were my app crash is due to the frame rate of the camera of my laptop, after testing it on other devices, I can affirm it work well.
